### PR TITLE
fix(telemetry): enable telemetry before starting the writer

### PR DESCRIPTION
**Pull request by mabdinur**
Currently we start the telemetry writer (which attempts to queue an app-started event) before we set `TelemetryWriter._enabled = True`.  
The impact of this issue is that we never send an app-started event. 
This bug was introduced in an unreleased change: https://github.com/DataDog/dd-trace-py/pull/3692. 
This bug was missed by ci due to how we setup the telemetry writer tests.

## Checklist
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: ddtrace/internal/telemetry/writer.py [hunk 1]
```diff
@@ -104,7 +108,7 @@ def periodic(self):
         if integrations:
             self._app_integrations_changed_event(integrations)
 
-        telemetry_requests = self._flush_events_queue()
+        telemetry_requests = self.flush_events_queue()
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:111]: I made `flush_events_queue` "public" to use it in a test. It seems like a helpful function to expose. I could revert this change since it's unreleased to this fix.

File: ddtrace/internal/telemetry/writer.py [hunk 2]
```diff
@@ -85,7 +89,7 @@ def _flush_integrations_queue(self):
             self._integrations_queue = []
         return integrations
 
-    def _flush_events_queue(self):
+    def flush_events_queue(self):
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:92]: I made flush_events_queue "public" to use it in a test. It seems like a helpful function to expose. I could revert this change since it's unreleased to this fix.

File: tests/tracer/telemetry/test_writer.py [hunk 3]
```diff
@@ -38,6 +38,8 @@ def mock_time():
 @pytest.fixture
 def telemetry_writer():
     telemetry_writer = TelemetryWriter(AGENT_URL)
+    # Enable the TelemetryWriter without queuing an app-started event
+    # and setting up exit hooks
     telemetry_writer._enabled = True
```
Comment (mabdinur) [inline tests/tracer/telemetry/test_writer.py:43]: This fixture masked the regression introduced in an earlier PR. I left a comment here to explain why setting `telemetry_writer._enabled = True` is necessary. Instead of leaving this comment I could try to refactor the tests instead.

File: tests/tracer/telemetry/test_telemetry.py [hunk 4]
```diff
@@ -1,37 +1,58 @@
...
-def test_enable():
-    telemetry.telemetry_writer.enable()
-    assert telemetry.telemetry_writer.status == ServiceStatus.RUNNING
+@pytest.fixture
+def telemetry_writer():
+    yield telemetry.telemetry_writer
+
+
+@pytest.fixture(autouse=True)
+def mock_send_request(telemetry_writer):
```
Comment (mabdinur) [inline tests/tracer/telemetry/test_telemetry.py:16]: Enabling/Disabling the telemetry writer queues and sends requests to the agent. We should mock the telemetry endpoint instead of sending events to the agent (the response does not matter, the response is validated in other tests).

**Comment by mabdinur** [comment]
@Mergifyio backport 1.2

File: ddtrace/internal/telemetry/writer.py [hunk 5]
```diff
@@ -268,8 +272,8 @@ def enable(self):
         if self.status == ServiceStatus.RUNNING:
             return
 
-        self.start()
         self._enabled = True
+        self.start()
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:276]: Here's the main fix. Set TelemetryWriter._enabled=true BEFORE starting the writer thread.

File: ddtrace/internal/telemetry/writer.py (hunk 2 above)
Comment (P403n1x87) [reply ddtrace/internal/telemetry/writer.py:88]: If this is unrelated, then I'd leave it out. Can we perhaps trigger this behaviour by calling `disable` instead?

Comment (mabdinur) [reply ddtrace/internal/telemetry/writer.py]: I updated the enable/disable tests so they do not depend on internal logic.

**Review by brettlangdon — APPROVED** [review]

**Comment by mabdinur** [comment]
Without this fix the telemetry writer will not send `app-started` events. I reverted my fix locally and the following test failed: 
```
================================================================================= FAILURES ==================================================================================
________________________________________________________________________________ test_enable ________________________________________________________________________________

telemetry_writer = TelemetryWriter(status=<ServiceStatus.RUNNING: 'running'>, _interval=60.0)

    def test_enable(telemetry_writer):
        telemetry_writer.enable()
        assert telemetry_writer.status == ServiceStatus.RUNNING
    
        # assert that calling enable twice does not raise an exception
        telemetry_writer.enable()
        assert telemetry_writer.status == ServiceStatus.RUNNING
    
        # send queued events
        telemetry_writer.periodic()
        # ensure an app-started telemetry request is sent
>       assert len(httpretty.latest_requests()) == 1
E       assert 0 == 1
E         +0
E         -1

tests/tracer/telemetry/test_telemetry.py:33: AssertionError
```

**Review by brettlangdon — APPROVED** [review]
